### PR TITLE
fix(ssr): make context providers work again

### PR DESCRIPTION
**Pull request by wjhsf**
## Does this pull request introduce a breaking change?

- 😮‍💨 No, it does not introduce a breaking change.
- 💔 Yes, it does introduce a breaking change.

## Does this pull request introduce an observable change?

- 🤞 No, it does not introduce an observable change.
- 🔬 Yes, it does include an observable change.

### CODE REVIEW

File: packages/@lwc/ssr-compiler/src/compile-template/index.ts [hunk 1]
```diff
@@ -32,6 +32,7 @@ const bExportTemplate = esTemplate`
         // These should be scoped to the "tmpl" function however, to avoid conflicts with other templates.
         let textContentBuffer = '';
         let didBufferTextContent = false;
+        const contextfulParent = instance;
```
Comment (nolanlawson) [inline packages/@lwc/ssr-compiler/src/compile-template/index.ts:35]: Let's add a comment explaining what this does.

Comment (nolanlawson) [reply packages/@lwc/ssr-compiler/src/compile-template/index.ts:35]:
```suggestion

        // Establishes a contextual relationship between two components for ContextProviders.
        // This variable will typically get overridden (shadowed) within slotted content.
        const contextfulParent = instance;
```

File: packages/@lwc/ssr-compiler/src/compile-template/index.ts [hunk 2]
```diff
@@ -55,7 +56,9 @@ const bExportTemplate = esTemplate`
         if (!isLightDom) {
             yield '</template>';
             if (shadowSlottedContent) {
-                yield* shadowSlottedContent();
+                // instance must be passed in; this is used to establish the contextful relationship
+                // between context provider (aka parent component) and context consumer (aka slotted content)
+                yield* shadowSlottedContent(instance);
```
Comment (nolanlawson) [inline packages/@lwc/ssr-compiler/src/compile-template/index.ts:61]: It's really weird that we pass in the `instance` here but the `contextualParent` when calling a light DOM slotted content generator... Something to investigate later. (Maybe we can break this with several layers of context providers / components.)

**Review by nolanlawson — APPROVED** [review]

File: packages/@lwc/ssr-compiler/src/compile-template/index.ts (hunk 2 above)
Comment (wjhsf) [reply packages/@lwc/ssr-compiler/src/compile-template/index.ts:61]: This is in the root `tmpl` function, so at this point in the code, `contextfulParent` will _always_ be `instance`. It doesn't matter which we use, but we can use `contextfulParent` for consistency.
```suggestion
                yield* shadowSlottedContent(contextfulParent);
```

Comment (nolanlawson) [reply packages/@lwc/ssr-compiler/src/compile-template/index.ts]: Whew okay, that makes sense!

**Review by divmain — APPROVED** [review]
Nice! LGTM

**Pull request merged**